### ISSUE TITLE
Calculating Data on "Unit Help" takes extraordinarily long time

### ISSUE THREAD

**Issue by tvleavitt** [triplea-game/triplea#7142]
### How can the problem be recreated?

Start TripleA. Start "Warcraft: War Heroes" Game. Go to Help/Unit Help. Minutes later, you still see "Calculating Data..." and nothing else can be done or interacted with. Game can't even be stopped via anything but "End Task" in Task Manager.

### If playing a prerelease, which version are you using?

Game Version: 2.1.20560, 0.1.3 (Map)

### If playing a prerelease, does this happen on the latest release?

- [X] yes
- [] no

### Is there anything else we should know?

Just FYI, monitoring memory usage in the console as initial deployment occurs, it looks like it is coming pretty close to using up all the heap space. This map has 35 "players". It's using 2GB of RAM, and 17.5% of my CPU and task manager now reports it as "Not Responding". I see 50-60 threads threads running in Performance Manager on Win10.

Heap utilization statistics [MB]
Used Memory: 1681
Free memory: 297
Total memory: 1979
Max memory: 1979

**Comment by DanVanAtta** [comment]
I've observed I'm unable to load unit help for WofW map, always get an out of heap space. Also, when heap space is low, the unit help renders in a very odd way: https://github.com/triplea-game/triplea/issues/6273

Loading up unit help, I notice we have a surge of `int[]` objects that appear to be the problem. Most images loaded via swing are kept as `byte[]`. This implies that fundamentally it is image rendering that is the problem.

It *looks* like we have a few issues:
1. The game does not know which unit types belong to which players. To get to this, the game looks at:
  - all units on map
  - all production frontier
  - all unit artworks
  If any of the above indicates a unit belongs to a player, then the game engine figures that unit type would belong to that player. This implies if you delete a neutral unit image, then that unit would simply no longer be available to be placed via edit (one would expect that to require an XML change).  

2. While looking for unit artworks, the game renders the image. Instead the game should check that the image file simply exists, it should not attempt to render it. This does result in a heavy initial load as the full set of units will be rendered, many of which are not present on the map and thus not available in cache. 

3. The unit help is rendered as HTML. It *appears* this causes a large number `int[]` objects to be created, I suspect these are to support the HTML images. It seems that the swing-HTML image rendering is perhaps not efficient and does not share with the already rendered (cached) images.  I do not actually know how much of an impact the HTML vs native swing rendering is having, this looks to be a difference why so much memory is consumed in unit help.


Solving (1) will require a long-term solution and some XML changes. I've opened this forum thread to start that conversation: https://forums.triplea-game.org/topic/2160/proposal-to-consolidate-unit-options-into-unitlist-make-explicit-which-units-are-available-to-which-players

(2) & (3) are more readily solved.

**Comment by Cernelius** [comment]
> 1. The game does not know which unit types belong to which players. To get to this, the game looks at:
> 
> * all units on map
> * all production frontier
> * all unit artworks

I always felt grabbing whatever unit images one can find in the skin to be a very unpolished solution. An extreme example of what this may cause you can find in "World At War" if you open the "Unit Help" thing and go to French. There you can see a blue-violet coloured "Marine" unit, which has nothing to do with the game or even the map, and it is just there because that player happens to have the same name as one of the programs asset's players, thus getting the image at this location:
`TripleA/assets/units/French`

Here it is an example of a user making a review of "World At War" with a number of inaccuracies, caused by the "Unit Help" table duping it into believing that a lot of units more were in the game (while they were units inexistent in the game, that the program was and is finding images for no reasons):
https://forums.triplea-game.org/topic/1659/made-a-promotional-video-to-grow-community/8
>their [French] marines are coloured differently, for some reason

>they [Anzac] do get heavy tanks

>[Anzac] and [incomprehensible to me] the carrier as well

>the Finns get heavy tanks

>the Romanians get pretty much everything [incomprehensible to me] the heavy tanks, same for all the other Axis

>heavy tanks. So they [Americans] get everything

All the quoted statements are wrong (there are other cases too).

**Comment by DanVanAtta** [comment]
Yeah, fully agree, looking for all available images is kinda backwards. Hence the engine needs to be 'told' which unit types belong to which players. The only complexity is ensuring completeness given possible capture rules.

**Comment by DanVanAtta** [comment]
#7249 should help a lot. If the performance is still sub-par, we can go further (use Swing instead of HTML, use a lazy-loaded tab view so that we only render one player at a time). After #7249 is merged I'll be curious how things have improved.

**Comment by Cernelius** [comment]
Personally, for a simple fix, I would rather eliminate the behaviour of searching for available unit images entirely (meaning having in unit help only units owned or available for purchase or both).

I know that, this way, you will not have information for other kind of available units, but, in my opinion, games having such advanced elements should take care to document them in notes and, anyway, once the unit is eventually actually available, you will have it in unit help too. Regardless, I think it makes sense unit help displaying only the units you currently have available. For example, this way, in World At War, you would have the "red dot" units in unit help only if you open it after the start of round 4. If you want to know ahead what are the abilities of the units that you will have, it makes sense that the notes should do this (and the notes of World At War do it fully and flawlessly, except only for the fact that they don't display all unit images, but you need this only once the units are actually on the board (at which point you have them in Unit Help)).

Long story short, out of this:
>- all units on map
>- all production frontier
>- all unit artworks

I suggest completely to eliminate the third item.

**Comment by DanVanAtta** [comment]
@Cernelius the forum thread is the best place to discuss unit-type to player associations

**Comment by tvleavitt** [comment]
So, I just loaded "Unit Help". It loads. Still takes a painfully long time, and the "Please wait" pop up disappears well before the screen actually loads, freezing the entire game in the process.

Loading the game a second time seems to have hung TripleA altogether.

**Comment by DanVanAtta** [comment]
This was drastically improved I think in 2.5. Clock time to open unit help is relatively short, under 5 seconds now on reasonably large maps (WaW). Resolving..

**Comment by tvleavitt** [comment]
This still takes a very long time on Warcraft War Heroes. You see "Processing data..." or somesuch, then nothing for well over 30 seconds while the entire game is frozen, then the Unit Help screen finally pops up (this on a 2018 era i5 system with 32GB RAM). Could this possibly be fixed by breaking this screen up into multiple tabs (one per player), or only loading part of the menu and adding a "View More" option?
